src/exodus/sigmawallet.h

```diff
@@ -0,0 +1,156 @@
+    CKeyID seedId;
+
+    MintPoolEntry();
+    MintPoolEntry(SigmaPublicKey const &key, CKeyID const &seedId);
+
+    bool operator==(MintPoolEntry const &) const;
+    bool operator!=(MintPoolEntry const &) const;
+
+    ADD_SERIALIZE_METHODS;
+
+    template<typename Stream, typename Operation>
+    void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion)
+    {
+        READWRITE(key);
+        READWRITE(seedId);
+    }
+};
+
+class SigmaWallet
+{
+protected:
+
+    typedef boost::multi_index_container<
+        MintPoolEntry,
+        boost::multi_index::indexed_by<
+            // Sequence
+            boost::multi_index::sequenced<>,
+            // Public Key index
+            boost::multi_index::hashed_unique<
+                boost::multi_index::member<MintPoolEntry, SigmaPublicKey, &MintPoolEntry::key>,
+                std::hash<SigmaPublicKey>
+            >
+        >
+    > MintPool;
+
+    std::string walletFile;
+    MintPool mintPool;
+    uint160 masterId;
+
+    static constexpr unsigned mintPoolCapacity = 20;
+
+public:
+    SigmaWallet();
+
+private:
+    void ReloadMasterKey();
+
+    // Generator
+private:
+    uint32_t GenerateNewSeed(CKeyID &seedId, uint512 &seed);
+    uint32_t GenerateSeed(CKeyID const &seedId, uint512 &seed);
+    uint32_t GetSeedIndex(CKeyID const &seedId);
+
+protected:
+    bool GeneratePrivateKey(uint512 const &seed, exodus::SigmaPrivateKey &coin);
+
+    // Mint updating
+public:
+    SigmaPrivateKey GeneratePrivateKey(CKeyID const &seedId);
+    std::pair<SigmaMint, SigmaPrivateKey> GenerateMint(
+        PropertyId propertyId,
+        SigmaDenomination denom,
+        boost::optional<CKeyID> seedId = boost::none);
+
+    void ClearMintsChainState();
+    bool TryRecoverMint(
+        SigmaMintId const &id,
+        SigmaMintChainState const &chainState);
+
+private:
+    bool SetMintSeedSeen(
+        MintPoolEntry const &mintPoolEntry,
+        PropertyId propertyId,
+        SigmaDenomination denomination,
+        exodus::SigmaMintChainState const &chainState,
+        uint256 const &spendTx = uint256());
+    SigmaMint UpdateMint(SigmaMintId const &, std::function<void(SigmaMint &)> const &);
+    void WriteMint(SigmaMintId const &id, SigmaMint const &entry);
+
+public:
+    SigmaMint UpdateMintChainstate(SigmaMintId const &id, SigmaMintChainState const &state);
+    SigmaMint UpdateMintSpendTx(SigmaMintId const &id, uint256 const &tx);
+
+    // Mint querying
+public:
+
+    bool HasMint(SigmaMintId const &id) const;
+    bool HasMint(secp_primitives::Scalar const &serial) const;
+    SigmaMint GetMint(SigmaMintId const &id) const;
+    SigmaMint GetMint(secp_primitives::Scalar const &serial) const;
+    SigmaMintId GetMintId(secp_primitives::Scalar const &serial) const;
+
+    template<
+        class OutIt,
+        typename std::enable_if<is_iterator<OutIt>::value>::type* = nullptr
+    > OutIt ListMints(OutIt it, bool unusedOnly, bool matureOnly) const
```

Comment:
IMO the two conditional parameters make code more hard to read from the caller site. Better to remove it and allow the caller site to filter by them self will looks better.